File: python-node/shape.py
import data, math, random

# Our nice sigmoid function
def sigmoid(x):
    return 1 / (1 + math.exp(-x))

# All the node classes
class ValueHolder:
    def getValue(self):
        raise NotImplementedError()

class Node(ValueHolder):
    pass

class InputNode(Node):
    def getValue(self):
        return self.value

class Link(ValueHolder):
    def __init__(self, inputNode):
        self.weight = random.uniform(-1, 1)
        self.inputNode = inputNode

    def getValue(self):
        return self.weight * self.inputNode.getValue()

class OutputNode(Node):
    def __init__(self, links):
        self.links = links

    def getValue(self):
        return sigmoid(sum([ link.getValue() for link in self.links ]))

# Our neural network class
class NeuralNetwork:
    # Create input and outputs on init
    def __init__(self, inputNodesCount, outputNodesCount):
        self.inputNodes = [ InputNode() for i in range(inputNodesCount) ]
        self.outputNodes = [ OutputNode([ Link(inputNode) for inputNode in self.inputNodes ]) for i in range(outputNodesCount) ]

    # To run the network update all the input value nodes then get value from output nodes
    def run(self, input):
        for i, value in enumerate(input):
            self.inputNodes[i].value = value
        return [ outputNode.getValue() for outputNode in self.outputNodes ]

    # To calculate the error subtract given value from wanted value and square that for each symbol for each training item
    def __error(self, trainingItems, symbols):
        errorSum = 0
        for item in trainingItems:
            result = self.run(item[0])
            correct = symbols[item[1]]
            errorSum += sum([ (result[i] - correct[i]) ** 2 for i in range(len(symbols)) ])
        return errorSum / len(trainingItems)

    # Adjust the weights random until the error value is less then the max error argument
    def train(self, trainingItems, symbols, maxError):
        trainingCycles = 0
        error = self.__error(trainingItems, symbols)
        while error > maxError:
            changes = []
            for outputNode in self.outputNodes:
                randomLink = outputNode.links[random.randint(0, len(outputNode.links) - 1)]
                changes.append({ 'link': randomLink, 'oldWeight': randomLink.weight })
                randomLink.weight += random.uniform(-1, 1) / 100

            newError = self.__error(trainingItems, symbols)
            if newError < error:
                error = newError
            else:
                for change in changes:
                    change['link'].weight = change['oldWeight']

            trainingCycles += 1
        return trainingCycles

# Create neural network train and test
if __name__ == '__main__':
    network = NeuralNetwork(data.inputDim, data.outputDim)

    print('Training...')
    trainingCycles = network.train(data.trainingSet, data.outputDict, data.maxError)
    print('Training done in', trainingCycles, 'cycles!')

    for item in data.testSet:
        result = network.run(item[0])
        correct = data.outputDict[item[1]]
        print(result, correct, (result[0] > result[1]) == (correct[0] > correct[1]) and 'PASSED' or 'FAILED')

File: native-matrix/shape.c
#include <stdio.h>
#include <stdlib.h>
#include "neural.h"

typedef struct DataItem {
    float input[9];
    float output[2];
} DataItem;

int main(void) {
    random_init();

    size_t networkLayers[] = { 9, 6, 2 };
    NeuralNetwork *network = neural_network_new(
        ACTIVATION_SIGMOID,
        networkLayers, sizeof(networkLayers) / sizeof(size_t)
    );

    // Train network
    DataItem trainItems[] = {
        {{
            1, 1, 1,
            1, 0, 1,
            1, 1, 1
        }, {1, 0}},
        {{
            0, 1, 0,
            1, 0, 1,
            0, 1, 0
        }, {1, 0}},
        {{
            0, 1, 0,
            1, 1, 1,
            0, 1, 0
        }, {0, 1}},
        {{
            1, 0, 1,
            0, 1, 0,
            1, 0, 1
        }, {0, 1}}
    };
    printf("Training...\n");
    size_t trainingCycles = neural_network_train(network, (float *)trainItems, sizeof(trainItems) / sizeof(DataItem), 0.005);
    printf("Training done in %zu cycles!\n", trainingCycles);

    // Test network
    DataItem testItems[] = {
        {{
            0, 1, 1,
            1, 0, 1,
            1, 1, 0
        }, {1, 0}},
        {{
            1, 0, 1,
            1, 0, 1,
            1, 1, 0
        }, {1, 0}},
        {{
            1, 0, 0,
            1, 1, 1,
            0, 0, 1
        }, {0, 1}},
        {{
            0, 0, 1,
            1, 1, 1,
            1, 0, 0
        }, {0, 1}},
        {{
            1, 0, 1,
            0, 0, 0,
            1, 0, 1
        }, {1, 0}},
        {{
            0, 1, 0,
            1, 0, 1,
            0, 1, 0
        }, {1, 0}},
        {{
            0, 0, 0,
            1, 1, 1,
            0, 0, 0
        }, {0, 1}},
        {{
            1, 0, 0,
            1, 1, 0,
            1, 0, 0
        }, {0, 1}}
    };
    for (size_t i = 0; i < (sizeof(testItems) / sizeof(DataItem)); i++) {
        DataItem *testItem = &testItems[i];
        float output[2];
        neural_network_run(network, testItem->input, output);
        printf("%f %f | %f %f | %s\n", output[0], output[1], testItem->output[0], testItem->output[1],
            output[0] > output[1] == testItem->output[0] > testItem->output[1] ? "PASSED" : "FAILED");
    }

    // Clean up
    neural_network_free(network);
    return EXIT_SUCCESS;
}
